python/pyarrow/tests/parquet/test_dataset.py

```diff
@@ -206,7 +206,7 @@ def test_filters_equivalency(tempdir, use_legacy_dataset):
     dataset = pq.ParquetDataset(
         base_path, filesystem=fs,
         filters=[('integer', '=', 1), ('string', '!=', 'b'),
-                 ('boolean', '==', True)],
+                 ('boolean', '==', 'True')],
```

Comment:
Was this change needed? (it feels it shouldn't be)

Reply:
@jorisvandenbossche Previously we allowed the implicit cast from boolean to string, but after this PR comparing a column of type string to a boolean scalar is no longer supported. Column "boolean" has type string because no explicit schema is provided and it isn't inferred as integral, which leaves string.

Making matters worse: we *can't* provide an explicit schema because `_ParquetDatasetV2` doesn't support that kwarg and `ParquetDataset` doesn't support a custom partitioning.

Reply:
Ah, yes, that makes sense. 
In the old ParquetDataset code, this works because we simply always try to convert the value of the partition to the type of the value in the expression (although more logically would be the other way around, though ...). So the string "True" gets cast to a bool and then compared (which actually might also mean this never worked with False, as `bool("False") is True` ...)

https://github.com/apache/arrow/blob/4086409e1b4cf4feac3b5c84060c69e6c7de898d/python/pyarrow/parquet.py#L951-L952

input value of the "expression" to the value in the partition dictionary. So even if the partitioning has a string "True", we will convert the `True` value to